src/parser.rs

```diff
@@ -1,185 +1,266 @@
-        let line = try!(self.read_string_line());
-        if line == "OK" {
-            Ok(Value::Okay)
-        } else {
-            Ok(Value::Status(line))
-        }
+pub fn parse_async<R>(reader: R) -> ValueFuture<R>
+where
+    R: AsyncRead + BufRead,
+{
+    ValueFuture {
+        reader: Some(reader),
+        state: AnyPartialState::default(),
+        remaining: Vec::new(),
     }
+}
 
-    fn parse_int(&mut self) -> RedisResult<Value> {
-        Ok(Value::Int(try!(self.read_int_line())))
-    }
+/// The internal redis response parser.
+pub struct Parser<T> {
+    reader: T,
+}
 
-    fn parse_data(&mut self) -> RedisResult<Value> {
-        let length = try!(self.read_int_line());
-        if length < 0 {
-            Ok(Value::Nil)
-        } else {
-            let data = try!(self.read(length as usize));
-            try!(self.expect_newline());
-            Ok(Value::Data(data))
-        }
+/// The parser can be used to parse redis responses into values.  Generally
+/// you normally do not use this directly as it's already done for you by
+/// the client but in some more complex situations it might be useful to be
+/// able to parse the redis responses.
+impl<'a, T: BufRead> Parser<T> {
+    /// Creates a new parser that parses the data behind the reader.  More
+    /// than one value can be behind the reader in which case the parser can
+    /// be invoked multiple times.  In other words: the stream does not have
+    /// to be terminated.
+    pub fn new(reader: T) -> Parser<T> {
+        Parser { reader: reader }
     }
 
-    fn parse_bulk(&mut self) -> RedisResult<Value> {
-        let length = try!(self.read_int_line());
-        if length < 0 {
-            Ok(Value::Nil)
-        } else {
-            let mut rv = vec![];
-            rv.reserve(length as usize);
-            for _ in 0..length {
-                rv.push(try!(self.parse_value()));
-            }
-            Ok(Value::Bulk(rv))
-        }
-    }
+    // public api
 
-    fn parse_error(&mut self) -> RedisResult<Value> {
-        let desc = "An error was signalled by the server";
-        let line = try!(self.read_string_line());
-        let mut pieces = line.splitn(2, ' ');
-        let kind = match pieces.next().unwrap() {
-            "ERR" => ErrorKind::ResponseError,
-            "EXECABORT" => ErrorKind::ExecAbortError,
-            "LOADING" => ErrorKind::BusyLoadingError,
-            "NOSCRIPT" => ErrorKind::NoScriptError,
-            code => {
-                fail!(make_extension_error(code, pieces.next()));
-            }
+    pub fn parse_value(&mut self) -> RedisResult<Value> {
+        let mut parser = ValueFuture {
+            reader: Some(&mut self.reader),
+            state: AnyPartialState::default(),
+            remaining: Vec::new(),
         };
-        match pieces.next() {
-            Some(detail) => fail!((kind, desc, detail.to_string())),
-            None => fail!((kind, desc)),
+        match parser.poll()? {
```

Comment:
I should mention that instead of calling the asynchronous parsing here, we could parse directly from the `Read` instance. The only drawback is that the parser needs to be instantiated twice (once for `I == PartialStream<easy::Stream<&[u8]>>` and once for `I == ReadStream<R>`.